diff --git a/compact_forms.js b/compact_forms.js
index 723d76d..6a1f189 100644
--- a/compact_forms.js
+++ b/compact_forms.js
@@ -18,7 +18,7 @@ $.fn.compactForm = function (stars) {
         return;
       }
       var $field = $('#' + $label.attr('for'), context);
-      if (!$field.length || !$field.is('input:text,input:password,textarea')) {
+      if (!$field.length || !$field.is('input:text,input:password,textarea,input[type=email]')) {
         return;
       }
       // Store the initial field value, in case the browser is going to
