Index: includes/bootstrap.inc
===================================================================
--- includes/bootstrap.inc	(revision 0)
+++ includes/bootstrap.inc	(working copy)
@@ -1062,7 +1062,7 @@
       break;
 
     case DRUPAL_BOOTSTRAP_PATH:
-      require_once './includes/path.inc';
+      require_once variable_get('path_inc', './includes/path.inc');
       // Initialize $_GET['q'] prior to loading modules and invoking hook_init().
       drupal_init_path();
       break;
Index: includes/theme.maintenance.inc
===================================================================
--- includes/theme.maintenance.inc	(revision 0)
+++ platform/includes/theme.maintenance.inc	(working copy)
@@ -22,7 +22,7 @@
     return;
   }
 
-  require_once './includes/path.inc';
+  require_once variable_get('path_inc', './includes/path.inc');
   require_once './includes/theme.inc';
   require_once './includes/common.inc';
   require_once './includes/unicode.inc';
